### PR TITLE
VACMS-13852: Creates EntityEvent Subscriber.

### LINKED ISSUE #13852: Create `ContentReleaseEntityEventSubscriber` service.
## Description

Following #13851 and followup tickets, #13854, and #13855, we should replace `Drupal\va_gov_build_trigger\EventSubscriber\EntityEventSubscriber` with a compact, well-written class that outsources its logical decisions to the `ContentReleaseEntityEventResponseResolver` service as described in #13851.

## Acceptance Criteria

- [x] A new `ContentReleaseEntityEventSubscriber` service has been created to replace the `EntityEventSubscriber` class.
- [x] Unit tests confirm agreement with the previous system.

### LINKED ISSUE #14495: Facility related content releases are not logged in most environments
## Description
As a developer I am unable to assess if a facility related status change or banner update has or should have triggered a content release.  This makes it impossible to assess if the logic for triggered content releases is trying to fire.
The shouldTrigger() is basically an environment check that prevents any of the other checks from firing and results in no logging.
image

## Acceptance Criteria
- [ ] If shouldTrigger() returns FALSE and the other checks return TRUE, then a log entry should be made to show that a content_release would have been triggered, even though none was.
- [ ] If shouldTrigger() returns TRUE, then existing logic should result in content release being queued.

### Team
Please check the team(s) that will do this work.

- [ ] `CMS Team`
- [ ] `Public Websites`
- [x] `Facilities`
- [ ] `User support`
- [ ] `Accelerated Publishing`

### PR DESCRIPTION
**Pull request by ndouglas**
## Description

Closes #13852.

Closes #14495 as well.

I removed the service declaration for the current entity event subscriber to see what (if anything) breaks.

I did a direct port of Cameron's tests for the current version of EntityEventSubscriber. After some work to adjust for different types, all of them pass... which makes me deeply suspicious. I've been careful, and have even done partial copies of these tests, but that still seems _really_ unexpectedly fortunate.

I think I'm going to add a couple more test suites to test from a few different angles to see if I can tease out any weird corner cases. I might add an extended test suite for facility/banner updates etc because I'm not sure unit tests can really account for all of the weirdness in Drupal.

## QA Steps  (more probably forthcoming)

- [x] Automated tests pass.
image

- On the [Tugboat Environment](https://cms-phixcqe7hcfnfp5oow2f7wlhsckyuqfi.ci.cms.va.gov/):
   - Banners
     - [x] create  and save a [banner](https://pr14337-phixcqe7hcfnfp5oow2f7wlhsckyuqfi.ci.cms.va.gov/node/add/banner) node as `draft`. Recent log messages should indicate that it _would not_ have triggered a content release.
     - [x] Open and save a `banner` node as `published`. Recent log messages should indicate that it _would_ have triggered a content release.
     - [x] Edit a published banner `banner` node as `draft`. Recent log messages should indicate that it _would NOT_ have triggered a content release.
     - [x] Edit a published banner `banner` node as `published`. Recent log messages should indicate that it _would_ have triggered a content release.
  - VAMC System Alert 
     - [x] create a [VAMC system alert](https://pr14337-phixcqe7hcfnfp5oow2f7wlhsckyuqfi.ci.cms.va.gov/node/add/full_width_banner_alert) and don't publish it.  Recent log messages should indicate that it _would NOT_ have triggered a content release.
     - [x] Edit a VAMC system alert to change its description and PUBLISH it.  Recent log messages should indicate that it _would_ have triggered a content release.
     - [x] Edit a VAMC system alert to change its description and leave it published.  Recent log messages should indicate that it _would_ have triggered a content release.
     - [x] Edit a VAMC system alert to change its description and UNPUBLISH it.  Recent log messages should indicate that it _would_ have triggered a content release.
     - [x] Edit a VAMC system alert to change its description and leave it unpublished.  Recent log messages should indicate that it _would NOT_ have triggered a content release.
  - Facility Status
     - [ ] Edit a VAMC facility and change its status and publish. Recent log messages should indicate that it _would_ have triggered a content release.
     - [ ] Edit a VAMC facility and change its status description and publish. Recent log messages should indicate that it _would_ have triggered a content release.
     - [ ] Edit a VAMC facility and change nothing and publish. Recent log messages should indicate no content release.

## Followup Items
- [x] replace "exception occurred" with "N/A" in the log messages.

### CODE REVIEW

File: tests/phpunit/va_gov_content_types/unit/ModuleTest.php
Comment (ndouglas) [inline tests/phpunit/va_gov_content_types/unit/ModuleTest.php:1]: This is an ugly test and probably not that valuable.

File: tests/phpunit/va_gov_content_release/unit/EventSubscriber/EntityEventSubscriberTest.php
[the remark this answers is not in the document]
Comment (swirtSJW) [reply tests/phpunit/va_gov_content_release/unit/EventSubscriber/EntityEventSubscriberTest.php:1]: Do we need to leave a human breadcrumb that changes to one needs to be made in the other so they remain clone-ish?

File: tests/phpunit/va_gov_content_types/unit/Traits/ContentReleaseTriggerTraitTest.php [hunk 1]
```diff
@@ -117,9 +126,19 @@ public function hasTriggeringChangesDataProvider() {
...
       }
-
-      $name = sprintf("%sisModerated, %sisCmPublished, %sisPublished, %sisArchived, %sisDraft, %swasPublished, %sexpected",
-        $isModerated ? '' : '!', $isCmPublished ? '' : '!', $isPublished ? '' : '!', $isArchived ? '' : '!', $isDraft ? '' : '!', $wasPublished ? '' : '!', $expected ? '' : '!');
+      if (!$expected || $isModerated || $isPublished || !(!$isModerated && $wasPublished && !$isPublished)) {
+        continue;
+      }
+      $name = sprintf(
+        "%sisModerated, %sisCmPublished, %sisPublished, %sisArchived, %sisDraft, %swasPublished, %sexpected",
+        $isModerated ? '' : '!',
```
Comment (swirtSJW) [inline tests/phpunit/va_gov_content_types/unit/Traits/ContentReleaseTriggerTraitTest.php:134]: It is probably because I am hungry, but this pattern confounds my brain at the moment.   

If something is true it results in empty, and False results in !   ?

**Review by swirtSJW — COMMENTED** [review]
Code is looking good/amazing.  Thanks for taking this all on.  A couple questions and comments.

File: tests/phpunit/va_gov_content_release/unit/EventSubscriber/EntityEventSubscriberTest.php
[the remark this answers is not in the document]
Comment (ndouglas) [reply tests/phpunit/va_gov_content_release/unit/EventSubscriber/EntityEventSubscriberTest.php:1]: Hoping to get this merged before that becomes an issue 😁

File: docroot/modules/custom/va_gov_content_release/src/Strategy/Plugin/StrategyPluginBase.php [hunk 2]
```diff
@@ -25,6 +25,8 @@ abstract class StrategyPluginBase extends PluginBase implements StrategyPluginIn
 
   /**
    * {@inheritDoc}
+   *
+   * @codeCoverageIgnore
```
[the remark this answers is not in the document]
Comment (ndouglas) [reply docroot/modules/custom/va_gov_content_release/src/Strategy/Plugin/StrategyPluginBase.php:29]: It's not needed and I might remove it. I'd love to have an idea of code coverage for Code That Actually Matters™. And it's really good to test constructors, because I've found it's painfully easy to create classes that you can't actually instantiate because I didn't `use` the right class, etc -- but that doesn't matter for code coverage anyway, since I do that stuff in functional tests.

But I'm not married to it. It's probably not worth the effort to exclude all of these things just to have a "cleaner" code coverage report.

File: docroot/modules/custom/va_gov_content_release/src/EntityEvent/Strategy/Resolver/Resolver.php [hunk 3]
```diff
@@ -39,8 +39,8 @@ public function getStrategyId() : string {
     $environment = $this->environmentDiscovery->getEnvironment();
     return match (TRUE) {
       $environment->isBrd() => static::STRATEGY_ON_DEMAND,
-      $environment->isTugboat() => static::STRATEGY_NEVER,
-      $environment->isLocalDev() => static::STRATEGY_NEVER,
+      $environment->isTugboat() => static::STRATEGY_VERBOSE_NEVER,
```
[the remark this answers is not in the document]
Comment (ndouglas) [reply docroot/modules/custom/va_gov_content_release/src/EntityEvent/Strategy/Resolver/Resolver.php:42]: I was staring at those constant names earlier trying to think of something better, heh.

File: tests/phpunit/va_gov_content_types/unit/Traits/ContentReleaseTriggerTraitTest.php (hunk 1 above)
Comment (ndouglas) [reply tests/phpunit/va_gov_content_types/unit/Traits/ContentReleaseTriggerTraitTest.php:134]: This is to help keep our brains from blowing up if/when a test fails.

It causes test failures to look like this:

```
8) tests\phpunit\va_gov_content_types\unit\Traits\ContentReleaseTriggerTraitTest::testHasTriggeringChanges with data set "isModerated, isCmPublished, isPublished, !isArchived, !isDraft, !wasPublished, !expected" (true, true, true, false, false, false, false)
Failed asserting that true matches expected false.

/var/www/html/docroot/vendor/phpunit/phpunit/src/Framework/Constraint/Equality/IsEqual.php:96
/var/www/html/tests/phpunit/va_gov_content_types/unit/Traits/ContentReleaseTriggerTraitTest.php:70
/var/www/html/docroot/vendor/phpunit/phpunit/src/Framework/TestResult.php:728
/var/www/html/docroot/vendor/phpunit/phpunit/src/Framework/TestSuite.php:684
/var/www/html/docroot/vendor/phpunit/phpunit/src/Framework/TestSuite.php:684
/var/www/html/docroot/vendor/phpunit/phpunit/src/TextUI/TestRunner.php:651
/var/www/html/docroot/vendor/phpunit/phpunit/src/TextUI/Command.php:144
/var/www/html/docroot/vendor/phpunit/phpunit/src/TextUI/Command.php:97
```

So instead of saying "okay, the, uh, the fifth value is a false here -- is it false for the other failed tests? well, some of them yeah, okay, now what was the fifth value?" and flipping back and forth between the failed test message and the test code and so forth and so on, you can do something a bit easier.

You can say, "okay, this is moderated, it's published, it wasn't published before... so it SHOULD trigger a content release. But !expected means we're not expecting it to. So the expectation is wrong and the test is broken."

Or for this one:

```
2) tests\phpunit\va_gov_content_types\unit\Traits\ContentReleaseTriggerTraitTest::testHasTriggeringChanges with data set "isModerated, !isCmPublished, !isPublished, !isArchived, isDraft, wasPublished, !expected" (true, false, false, false, true, true, false)
Failed asserting that true matches expected false.

/var/www/html/docroot/vendor/phpunit/phpunit/src/Framework/Constraint/Equality/IsEqual.php:96
/var/www/html/tests/phpunit/va_gov_content_types/unit/Traits/ContentReleaseTriggerTraitTest.php:70
/var/www/html/docroot/vendor/phpunit/phpunit/src/Framework/TestResult.php:728
/var/www/html/docroot/vendor/phpunit/phpunit/src/Framework/TestSuite.php:684
/var/www/html/docroot/vendor/phpunit/phpunit/src/Framework/TestSuite.php:684
/var/www/html/docroot/vendor/phpunit/phpunit/src/TextUI/TestRunner.php:651
/var/www/html/docroot/vendor/phpunit/phpunit/src/TextUI/Command.php:144
/var/www/html/docroot/vendor/phpunit/phpunit/src/TextUI/Command.php:97
```

You can say "okay, it's moderated, it's not published, it's a draft, but  it triggered a content release. It shouldn't have, so !expected is correct. So there's something wrong with the the code under test. One of the conditions is returning true when it should return false." And it still might break your brain to debug it from there, but at least you're not fighting to decode the test error message _as much_.

That said, it's a really ugly and opaque line, and I don't mind summarizing this justification and sticking it in there as a comment 🙂

Comment (ndouglas) [reply tests/phpunit/va_gov_content_types/unit/Traits/ContentReleaseTriggerTraitTest.php:134]: Or I might be able to build it in the test itself, and print error messages like "This node is moderated, not published, a draft... but it triggered a content release." Which is at least more readable and might directly express the thinking that you have to do to parse the message 🤷🏻‍♂️

File: docroot/modules/custom/va_gov_content_release/src/Strategy/Plugin/StrategyPluginBase.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (swirtSJW) [reply docroot/modules/custom/va_gov_content_release/src/Strategy/Plugin/StrategyPluginBase.php:29]: That makes sense.  I thought maybe it was throwing an error or something that we were trying to avoid.

File: tests/phpunit/va_gov_content_types/unit/Traits/ContentReleaseTriggerTraitTest.php (hunk 1 above)
Comment (swirtSJW) [reply tests/phpunit/va_gov_content_types/unit/Traits/ContentReleaseTriggerTraitTest.php:131]: I get it now and makes sense.  I like it, even though it made me feel stupid ;)

**Comment by swirtSJW** [comment]
Related comment from the other issue https://github.com/department-of-veterans-affairs/va.gov-cms/issues/14495#issuecomment-1653717335

**Comment by swirtSJW** [comment]
@ndouglas While testing the save of a system full width banner, I notices this in the log message.  Do we want these showing as exceptions.  They are really just fields or properties that can not be assessed on this particular content type.

image

**Comment by ndouglas** [comment]
@swirtSJW Yeah, that's intentional on my part (though debatable). I wanted to make the logic of retrieving and displaying these details as simple as possible, to avoid having the same logic (e.g. `if ($isModerated) { $result['isDraft'] = $this->isDraft; }`) in multiple places. It might be better if I say "N/A" for each of those values.

**Comment by swirtSJW** [comment]
> It might be better if I say "N/A" for each of those values. That's really what is implied there.

That version seems less alarming and more informative. Or "Not available"

**Review by swirtSJW — APPROVED** [review]
Banner create: Pass
image

Banner publish:  Pass
image

System banner : Pass
image

VAMC Facility: passed 
image

This all tests out for me @ndouglas 

Really nice work.  Thank you for undertaking it.

**Review by swirtSJW — APPROVED** [review]
